internal/servers/worker/internal/metric/proxy_ws_test.go

```diff
@@ -0,0 +1,419 @@
+				require.EqualError(t, err, tt.expErrStr)
+				require.False(t, tpc.addCalled)
+				require.Equal(t, 0, n)
+				return
+			}
+
+			require.Equal(t, 1, n)
+			require.True(t, tpc.addCalled)
+		})
+	}
+}
+
+func TestWsConnWrite(t *testing.T) {
+	tests := []struct {
+		name      string
+		tc        *testConn
+		expErr    bool
+		expErrStr string
+	}{
+		{
+			name:      "underlying write err",
+			tc:        &testConn{t: t, writeErr: fmt.Errorf("oops!")},
+			expErr:    true,
+			expErrStr: "oops!",
+		},
+		{
+			name:   "no underlying write err",
+			tc:     &testConn{t: t, writeErr: nil},
+			expErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tpc := &testPrometheusCounter{t: t}
+			ogBytesSent := bytesSent
+			bytesSent = tpc
+			t.Cleanup(func() { bytesSent = ogBytesSent })
+
+			wsConn := &wsConn{Once: &sync.Once{}, Conn: tt.tc}
+
+			_, err := wsConn.Write([]byte{})
+			require.True(t, tt.tc.writeCalled) // Underlying Write is always called.
+			if tt.expErr {
+				require.EqualError(t, err, tt.expErrStr)
+				require.False(t, tpc.addCalled)
+				return
+			}
+
+			require.True(t, tpc.addCalled)
+		})
+	}
+}
+
+func TestWsConnClose(t *testing.T) {
+	tests := []struct {
+		name           string
+		tc             *testConn
+		closeCallCount int
+		expErr         bool
+		expErrStr      string
+	}{
+		{
+			name:           "underlying close err",
+			tc:             &testConn{t: t, closeErr: fmt.Errorf("oops!")},
+			closeCallCount: 1,
+			expErr:         true,
+			expErrStr:      "oops!",
+		},
+		{
+			name:           "no underlying close err",
+			tc:             &testConn{t: t, closeErr: nil},
+			closeCallCount: 1,
+			expErr:         false,
+		},
+		{
+			name:           "underlying close err, multiple calls",
+			tc:             &testConn{t: t, closeErr: fmt.Errorf("oops!")},
+			closeCallCount: 10,
+			expErr:         true,
+			expErrStr:      "oops!",
+		},
+		{
+			name:           "no underlying close err, multiple calls",
+			tc:             &testConn{t: t, closeErr: nil},
+			closeCallCount: 10,
+			expErr:         false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tpg := &testPrometheusGauge{t: t}
+			ogConnCount := activeConnCount
+			activeConnCount = tpg
+			t.Cleanup(func() { activeConnCount = ogConnCount })
+
+			wsConn := &wsConn{Once: &sync.Once{}, Conn: tt.tc}
+			for i := 1; i <= tt.closeCallCount; i++ {
+				err := wsConn.Close()
+				require.Equal(t, i, tt.tc.closeCalledN)
+				if tt.expErr {
+					require.EqualError(t, err, tt.expErrStr)
+				}
+
+				require.Equal(t, 1, tpg.decCalledN)
+				require.Equal(t, 0, tpg.incCalledN)
+			}
+		})
+	}
+}
+
+type testHttpHijacker struct {
+	// Mocks the underlying http.Hijacker in metricsWrapper.
+	http.ResponseWriter
+	t             *testing.T
+	retConn       net.Conn
+	retErr        error
+	hijackCalledN int
+}
+
+func (h *testHttpHijacker) Hijack() (net.Conn, *bufio.ReadWriter, error) {
```

Comment:
> Does the testHttpHijacker you wrote here behave close to the one used in reality?

This `testHttpHijacker` returns a connection along with a newly created `bufio.ReadWriter`, just like the real deal. To that extent, I'd say that it does behave like the real one.

> I think it's possible for the bufio.ReadWriter to have a non empty reader at the time Hijack returns.

This is correct - There can be data in the buffer.

> Would that impact our code at all and should that be covered in our tests?

I don't think so because we're using the incoming reader directly.

We could write a test where our test hijacker returns some data in the reader and we could then verify that the data is present in the ReadWriter we create and return?